### DISCUSSION TITLE
Problems with custom structure and structure field UtcTime

### DISCUSSION THREAD

**Discussion by johannwesely** [eclipse-milo/milo#1345] [General]
Hello,
 I get the following error message in uaexpert when using a custom structure with an utctime field:
![Screenshot 2024-11-25 162750](https://github.com/user-attachments/assets/94a3bd38-a115-4071-a051-619dbb07ed27)

 I already use other custom structures wich work fine. Im running on Milo/dev1.0
here is the code of my customStruct:

```
public class DT_ProcessingTimesDataType extends AbsStructureSubClass implements ToxUaStructure {

	public static final ExpandedNodeId TYPE_ID = ExpandedNodeId.parse(String.format("nsu=%s;s=%s", ToxNamespace.Name, "Types/DataTypes/BaseDataType/Structure/DT_ProcessingTimesDataType"));
	public static final ExpandedNodeId BINARY_ENCODING_ID = ExpandedNodeId.parse(String.format("nsu=%s;s=%s", ToxNamespace.Name, "Types/DataTypes/BaseDataType/Structure/DT_ProcessingTimesDataType/.BinaryEncoding"));
	//
	public static final String F1 = "StartTime";
	public static final String F2 = "StopTime";
	public static final String F3 = "AcquisitionDuration";
	public static final String F4 = "ProcessingDuration";
	private Instant startTime = Instant.now();
	private Instant stopTime = Instant.now();
	private Duration acquisitionDuration = Duration.ZERO;
	private Duration processingDuration = Duration.ZERO;

	public DT_ProcessingTimesDataType(Instant startTime, Instant stopTime, Duration aquisitionDuration, Duration processingDuration) {
		this.startTime = startTime;
		this.stopTime = stopTime;
		this.acquisitionDuration = aquisitionDuration;
		this.processingDuration = processingDuration;
	}

	public DT_ProcessingTimesDataType() {
		this(Instant.now(), Instant.now(), Duration.ZERO, Duration.ZERO);
	}

	

	@Override
	public ExpandedNodeId getTypeId() {
		return TYPE_ID;
	}

	public ExpandedNodeId getBinaryEncodingId() {
		return BINARY_ENCODING_ID;
	}

	@Override
	public StructureField[] getStructureField(OpcUaServer server) throws Exception {

		return new StructureField[] {getUtcTimeStructureField(F1, server), getUtcTimeStructureField(F2, server), getDurationStructureField(F3, server), getDurationStructureField(F4, server)};

	}

	@Override
	public BinaryDataTypeCodec getCodec() {
		return BinaryDataTypeCodec.from(new DT_ProcessingTimesDataType.Codec());
	}

	public Instant getStartTime() {
		return startTime;
	}

	public void setStartTime(Instant startTime) {
		this.startTime = startTime;
	}

	public Instant getStopTime() {
		return stopTime;
	}

	public void setStopTime(Instant stopTime) {
		this.stopTime = stopTime;
	}

	public Duration getAcquisitionDuration() {
		return acquisitionDuration;
	}

	public void setAcquisitionDuration(Duration acquisitionDuration) {
		this.acquisitionDuration = acquisitionDuration;
	}

	public Duration getProcessingDuration() {
		return processingDuration;
	}

	public void setProcessingDuration(Duration processingDuration) {
		this.processingDuration = processingDuration;
	}



	public static class Codec extends GenericDataTypeCodec<DT_ProcessingTimesDataType> {
		@Override
		public Class<DT_ProcessingTimesDataType> getType() {
			return DT_ProcessingTimesDataType.class;
		}

		@Override
		public DT_ProcessingTimesDataType decodeType(EncodingContext context, UaDecoder decoder) throws UaSerializationException {
			Instant startTime = decoder.decodeDateTime(F1).getJavaInstant();
			Instant stopTime = decoder.decodeDateTime(F2).getJavaInstant();
			Duration acquisitionTime = Duration.ofMillis(decoder.decodeInt64(F3));
			Duration processingTime = Duration.ofMillis(decoder.decodeInt64(F4));
			return new DT_ProcessingTimesDataType(startTime, stopTime, acquisitionTime, processingTime);
		}

		@Override
		public void encodeType(EncodingContext context, UaEncoder encoder, DT_ProcessingTimesDataType value) throws UaSerializationException {
			encoder.encodeDateTime(F1, new DateTime(value.startTime));
			encoder.encodeDateTime(F2, new DateTime(value.stopTime));
			encoder.encodeInt64(F3, value.acquisitionDuration.toMillis());
			encoder.encodeInt64(F4, value.processingDuration.toMillis());
		}

	}

}

public StructureField getUtcTimeStructureField(String field, OpcUaServer server) {
		return new StructureField(field, LocalizedText.NULL_VALUE, Identifiers.UtcTime, ValueRanks.Scalar, null, uint(0), false);
	}

public StructureField getDurationStructureField(String field, OpcUaServer server) {
		return new StructureField(field, LocalizedText.NULL_VALUE, Identifiers.Duration, ValueRanks.Scalar, null, uint(0), false);
	}
```

Do you have any Idea what could go wrong?  I also register customStruct in dictionaryManager.registerStructure(structureClassName, dataTypeId, binaryEncodingId, uaStructure.getCodec(), description); and getNodeContext().getServer().getDataTypeManager().registerType(dataTypeId, uaStructure.getCodec(), binaryEncodingId, null, null);

**Comment by kevinherron** [comment]
Does it work if you define the field's type as DateTime instead? i.e. use `Identifiers.DateTime` rather than `Identifiers.UtcTime`? Just trying to narrow down whether it's related specifically to using this simple/alias type or if it's a broader issue somehow.

**Reply by kevinherron** [reply]
Hmm, from the UaExpert log it seems that it's trying to use legacy datatype dictionary mechanism rather than the DataTypeDefinition attribute... is that intentional? Is your server also still trying to build a datatype dictionary and expose it to clients?

**Comment by johannwesely** [comment]
Hello I changed the fields Type to DateTime then i got the same error code for duration. 
![Screenshot 2024-11-26 154303](https://github.com/user-attachments/assets/62e67060-1679-4d18-b781-b11e4754a517)
then i changed duration also to int64 and now i can see the struct in uaexpert (But of course with the wrong data types)
![2024-11-26 15_51_35-Greenshot](https://github.com/user-attachments/assets/f8096631-4321-4ca7-ba61-4c9627c05363)
so there seems to be a problem with utctime and duration as well. How can i check the things you mention above? ( legacy datatype dictionary ,expose to client) maybe i missed a line where i do this things without knowing.

**Reply by kevinherron** [reply]
I suspect you're doing this as part of defining your struct: https://github.com/eclipse-milo/milo/blob/3824869bfb842f666c9b995b6a9a27a00ab32b21/milo-examples/server-examples/src/main/java/org/eclipse/milo/examples/server/ExampleNamespace.java#L927-L944

I would avoid this, and avoid maintaining a `DataTypeDictionaryManager` instance at all. Support for these legacy datatype dictionaries is unlikely to make it into 1.0...

**Reply by kevinherron** [reply]
`UtcTime` and `Duration` are both what the spec calls "simple" or "alias" types - they are just aliases of some other concrete type - and it seems the datatype dictionary stuff does not support those correctly.

**Reply by johannwesely** [reply]
you are right i do this. But if i dont register the custom struct i get 
![image](https://github.com/user-attachments/assets/b7acaeca-454a-41b6-8387-df89b4df831c)
so what is the new mechanism to make custom structs available ?

```
protected void registerToxStructType(ToxUaStructure uaStructure) throws Exception {
-->		//BinaryDataTypeDictionaryManager dictManager = dictionaryManager.get(getServer().getNamespaceTable().getIndex(uaStructure.getBinaryEncodingId().getNamespaceUri()).intValue());

		NodeId dataTypeId = uaStructure.getTypeId().toNodeIdOrThrow(getServer().getNamespaceTable());

		NodeId binaryEncodingId = uaStructure.getBinaryEncodingId().toNodeIdOrThrow(getServer().getNamespaceTable());

		String structureClassName = uaStructure.getClass().getSimpleName();

		StructureField[] fields = uaStructure.getStructureField(getServer());

		StructureDefinition definition = new StructureDefinition(binaryEncodingId, Identifiers.Structure, StructureType.Structure, fields);

		StructureDescription description = new StructureDescription(dataTypeId, new QualifiedName(getNamespaceIndex(), structureClassName), definition);

	-->	//dictManager.registerStructure(structureClassName, dataTypeId, binaryEncodingId, uaStructure.getCodec(), description);

		// Add a custom DataTypeNode with a SubtypeOf reference to Structure
		UaDataTypeNode dataTypeNode = new UaDataTypeNode(getNodeContext(), dataTypeId, newQualifiedName(structureClassName), LocalizedText.english(structureClassName), LocalizedText.NULL_VALUE, uint(0), uint(0), false);

		dataTypeNode.addReference(new Reference(dataTypeId, NodeIds.HasSubtype, NodeIds.Structure.expanded(), Reference.Direction.INVERSE));

		getNodeManager().addNode(dataTypeNode);

		getNodeContext().getServer().getDataTypeManager().registerType(dataTypeId, uaStructure.getCodec(), binaryEncodingId, null, null);


	}
```

**Reply by kevinherron** [reply]
Everything preceding the comment in the method I just linked to; creating the `StructureDefinition` to set on the `DataTypeDefinition` attribute, etc...

Starting in OPC UA 1.04 the XML datatype dictionary was deprecated and the DataTypeDefinition attribute was introduced. Reading the DataTypeDefinition attribute is the only thing a Client needs to do to learn about a custom structure, and the only thing you need to do to publish it to a Client (though there is other stuff you must do with codecs etc.. to actually be able to use it).

**Comment by johannwesely** [comment]
Hi FYI,
i removed the DictionaryManager but first i missed the line dataTypeNode.setDataTypeDefinition(definition);  now i can read the structure with ua expert.
